model.py

```diff
@@ -137,3 +192,246 @@ def forward(self, inputs: Dict[str, th.Tensor]) -> th.Tensor:
+        sx, sy, sz = [s[..., i] for i in range(3)]
+
+        # NOTE(ycho):
+        # Represents Rz@Ry@Rx matrix,
+        # Which is equivalent to:
+        # rotate_yaw(rotate_pitch(rotate_roll(point))).
+        r = th.stack([
+            cy * cz,
+            (sx * sy * cz) - (cx * sz),
+            (cx * sy * cz) + (sx * sz),
+            cy * sz,
+            (sx * sy * sz) + (cx * cz),
+            (cx * sy * sz) - (sx * cz),
+            -sy,
+            (sx * cy),
+            (cx * cy)], dim=-1)
+
+        r = einops.rearrange(r, '... (i o) -> ... i o', i=3, o=3)
+        return r
+
+
+class MotionPredictor(nn.Module):
+    """DSR-Net Motion Predictor.
+
+    Predicts scene flow from scene representation and action embeddings.
+    Uses euler angle parameterization internally.
+    """
+
+    def __init__(self, use_action: bool, num_objects: int = 5,
+                 num_params: int = 6):
+        """
+        Args:
+            use_action: Whether `action` embeddings will be provided as input.
+                Otherwise, will generate a zero-filled tensor with same shape.
+            num_objects: Maximum number of object instances to predict.
+            num_params: The number of parameters for the transform encoding.
+                By default, uses 6 = (3 for translation, 3 for rotation).
+        """
+        super().__init__()
+        # NOTE(ycho): `use_action` is only `False`
+        # while we don't have ActionEmbedding() implemented.
+        self.use_action: bool = use_action
+        self.num_objects: int = num_objects
+        self.num_params: int = num_params
+
+        self.conv3d0 = Conv3D(8 + 8, 8, 2)
+        self.conv3d1 = Conv3D(8 + 8, 16, 2)
+        self.conv3d2 = nn.Sequential(*[
+            Conv3D(i, o, s) for
+            (i, o), s in zip(pairwise(
+                [16 + 16, 32, 32, 32, 64]), (2, 1, 1, 1))])
+
+        self.conv3d3 = nn.Sequential(
+            Conv3D(64, 128, 2),
+            Conv3D(128, 128, 2),
+            nn.Conv3d(128, 128, kernel_size=(4, 4, 2))
+        )
+
+        # 8, 64, 64, 64, 64 (,8)
+        act1_channels = (8, 64, 64, 64, 64)
+        act1_layers = []
+        for i, (c_in, c_out) in epairwise(act1_channels):
+            stride = (2 if i == 0 else 1)
+            act1_layers.append(Conv2D(c_in, c_out, stride))
+        self.action1 = nn.Sequential(*act1_layers)
+        self.action1e = Conv2D(64, 8, 1)
+
+        # 64, 128, 128, 128, 128 (,16)
+        act2_channels = (64, 128, 128, 128, 128)
+        act2_layers = []
+        for i, (c_in, c_out) in epairwise(act2_channels):
+            stride = (2 if i == 0 else 1)
+            act2_layers.append(Conv2D(c_in, c_out, stride))
+        self.action2 = nn.Sequential(*act2_layers)
+        self.action2e = Conv2D(128, 16, 1)
+
+        # MLP
+        # NOTE(ycho): last "object" is reserved for background.
+        # WARN(ycho): 5x512 in the paper, 4x512 hidden layers in the author's
+        # code.
```

Comment:
> After five fully connected layers with 512 hidden units,

Doesn't that mean 128->512->512->512->512->6(K-1)? The sentence is a little bit unclear, though.

Reply:
perhaps. I'll just go with the author's code